File: app/src/main/java/com/jawnnypoo/openmeh/services/PostReminderService.java
package com.jawnnypoo.openmeh.services;

import android.app.IntentService;
import android.content.Intent;
import android.graphics.Bitmap;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.Target;
import com.jawnnypoo.openmeh.api.MehClient;
import com.jawnnypoo.openmeh.api.MehResponse;
import com.jawnnypoo.openmeh.data.Deal;
import com.jawnnypoo.openmeh.util.MehNotificationManager;
import com.jawnnypoo.openmeh.util.MehPreferencesManager;

import java.io.IOException;
import java.util.concurrent.ExecutionException;

import retrofit2.Response;
import timber.log.Timber;

/**
 * Service that pretty much just posts a notification then goes away
 * Created by Jawn on 4/20/2015.
 */
public class PostReminderService extends IntentService {

    private static final String TAG = PostReminderService.class.getSimpleName();

    public PostReminderService() {
        super(TAG);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        postNotification();
    }

    private void postNotification() {
        if (!MehPreferencesManager.getNotificationsPreference(getApplicationContext())) {
            //Notifications disabled, go away
            return;
        }
        Response<MehResponse> response = null;
        try {
            response = MehClient.instance().getMeh().execute();
        } catch (IOException error) {
            Timber.e(error.toString());
        }
        if (response == null || !response.isSuccess() || response.body() == null) {
            Timber.e("Response was null. Will not notify");
            return;
        }


        Deal deal = response.body().getDeal();
        Bitmap icon = null;
        //Shoot for the highest resolution
        //http://graphicdesign.stackexchange.com/questions/15776/issues-with-creating-a-hi-res-large-icon-for-android-notifications-in-jelly-bean
        try {
            icon = Glide.with(getApplicationContext())
                    .load(deal.getPhotos().get(0))
                    .asBitmap()
                    .centerCrop()
                    .into(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
                    .get();
        } catch (ExecutionException | InterruptedException e) {
            Timber.e(e.toString());
        }
        MehNotificationManager.postDailyNotification(getApplicationContext(), response.body(), icon);
    }
}

File: app/src/main/java/com/jawnnypoo/openmeh/service/PostReminderService.kt
package com.jawnnypoo.openmeh.service

import android.app.IntentService
import android.content.Intent
import android.graphics.Bitmap
import com.bumptech.glide.Glide
import com.jawnnypoo.openmeh.App
import com.jawnnypoo.openmeh.shared.response.MehResponse
import com.jawnnypoo.openmeh.util.MehNotificationManager
import com.jawnnypoo.openmeh.util.Prefs
import timber.log.Timber
import java.lang.Exception

/**
 * Service that pretty much just posts a notification then goes away
 */
class PostReminderService : IntentService(PostReminderService.TAG) {

    companion object {
        private val TAG = PostReminderService::class.java.simpleName
    }

    override fun onHandleIntent(intent: Intent?) {
        postNotification()
    }

    private fun postNotification() {
        if (!Prefs.getNotificationsPreference(applicationContext)) {
            //Notifications disabled, go away
            return
        }
        var response: MehResponse? = null
        try {
            response = App.get().meh.getMeh()
                    .blockingGet()
        } catch (error: Exception) {
            Timber.e(error)
        }

        if (response == null || response.deal == null) {
            Timber.e("Response was null or deal was null. Will not notify")
            return
        }


        val deal = response.deal
        var icon: Bitmap? = null
        //Shoot for the highest resolution
        //http://graphicdesign.stackexchange.com/questions/15776/issues-with-creating-a-hi-res-large-icon-for-android-notifications-in-jelly-bean
        try {
            icon = Glide.with(applicationContext)
                    .asBitmap()
                    .load(deal?.photos?.firstOrNull())
                    .submit()
                    .get()
        } catch (e: Exception) {
            Timber.e(e)
        }

        MehNotificationManager.postDailyNotification(applicationContext, response, icon)
    }
}
